### PR TITLE
487 tiller error frontend

### PR DESCRIPTION
**Pull request by migmartri**
Continuation of https://github.com/kubeapps/kubeapps/issues/487

This issue was affecting the client code as well since we retrieve the resources from the releases directly from tiller.

This means that some filtering was necessary in the client too.

The code that performs that filter is really simple, the most difficult part has been to test the functionality, which requires defining/mocking some browser specific APIs that are not available in nodejs (by defaul) like websockets or localstorage. Both of them used by this component.

Ideally, we should extend these tests to leverage the websockets-mock library instead of limiting us to know that some socket instances were created.

### CODE REVIEW

File: dashboard/package.json [hunk 1]
```diff
@@ -51,8 +51,15 @@
     "lint-fix": "tslint -c ./tslint.ci.json --fix 'src/**/*.{ts,tsx}'"
   },
   "lint-staged": {
-    "*.{ts,tsx}": ["tslint -c ./tslint.ci.json --fix", "prettier --write", "git add"],
-    "*.{js,json,css,md}": ["prettier --write", "git add"]
+    "*.{ts,tsx}": [
```
Comment (migmartri) [inline dashboard/package.json]: I honestly don't know when or why this has changed, I guess that it is part of the `yarn add` ?

File: dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const manifest = `
```
Comment (andresmgot) [inline dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: I am not a big fan of static files for unit testing because:

 - They are not directly related to a specific unit test. What parts are you interested in? where are they used?
 - They are not reusable. You cannot extend/modify a file since it may break an existing unit test. This lead to the growth of these kind of files which become very difficult to manage.

I'd rather move this kind of manifest to a factory method (or just static vars) that returns only the piece of manifest you are interested in.

File: dashboard/src/components/AppView/AppView.tsx [hunk 3]
```diff
@@ -14,12 +14,13 @@ import "./AppView.css";
...
 import ServiceTable from "./ServiceTable";
 
-interface IAppViewProps {
+export interface IAppViewProps {
   namespace: string;
   releaseName: string;
   app: hapi.release.Release;
-  error: Error;
-  deleteError: Error;
+  // TODO(miguel) how to make optional props? I tried adding error? but the container complains
```
Comment (andresmgot) [inline dashboard/src/components/AppView/AppView.tsx]: Indeed that's weird. I don't know why the signature `error: Error | undefined` is different from `error?: Error`. It's even more weird because in the container we are instantiating the view props with `error: apps.error` where `apps.error` is an `error?: Error`.

Anyway I think that defining it as `Error | undefined` is a good solution.

File: dashboard/src/setupTests.ts [hunk 4]
```diff
@@ -2,5 +2,16 @@ import "raf/polyfill"; // polyfill for requestAnimationFrame
 
 import { configure } from "enzyme";
 import * as Adapter from "enzyme-adapter-react-16";
+import { WebSocket } from "mock-socket";
 
 configure({ adapter: new Adapter() });
+
+// Mock browser specific APIs like localstorage or Websocket
```
Comment (andresmgot) [inline dashboard/src/setupTests.ts]: nice find

File: dashboard/package.json [hunk 5]
```diff
@@ -9,10 +9,12 @@
     "@types/qs": "^6.5.1",
     "@types/react-jsonschema-form": "^1.0.4",
     "@types/react-select": "^1.2.6",
+    "@types/ws": "^6.0.0",
     "axios": "^0.17.1",
     "enzyme": "^3.3.0",
     "enzyme-adapter-react-16": "^1.1.1",
     "js-yaml": "^3.10.0",
+    "mock-socket": "^8.0.2",
```
Comment (andresmgot) [inline dashboard/package.json]: isn't this a `devDependency`?

File: dashboard/src/components/AppView/AppView.test.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  });
+
+  describe("State initialization", () => {
+    /*
+      The imported manifest contains one deployment, one service, one config map and some bogus manifests.
+      We only set websockets for deployment and services
+    */
+    it("sets a list of web sockets for its deployments and services", () => {
+      const wrapper = shallow(<AppViewComponent {...validProps} />);
+      wrapper.setProps(validProps);
```
[the remark this answers is not in the document]
Comment (migmartri) [reply dashboard/src/components/AppView/AppView.test.tsx]: We need to re-set props so componentWillReceiveProps gets triggered. it is not clear though so I have added a comment, thanks for the pointer.

In future refactorings I'd like to extract the state initialization to a function that is called on component mount and move away from the deprecated `componentwillreceiveprops`

File: dashboard/src/components/AppView/AppView.tsx (hunk 3 above)
Comment (migmartri) [reply dashboard/src/components/AppView/AppView.tsx]: Thanks for checking!

File: dashboard/package.json (hunk 5 above)
Comment (migmartri) [reply dashboard/package.json]: It was not working for me in dev mode maybe because tests load a different env. 

UPDATE: Just tested it and it works as dev dependency! Good catch @andresmgot! Thanks!

File: dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx (hunk 2 above)
Comment (migmartri) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: I do not think that I fully understand the concern so I'd like to know more.

> They are not directly related to a specific unit test. What parts are you interested in? where are they used?

Right, but they are related to a specific component and hence is inside component/testdata. Similarly you can find it in gocode, it might not be idiomatic here though. 

The reason I use a plain manifest is to test the convoluted componentWillReceiveProps which parses the yaml, splits resources by the separator and puts them in the different buckets.

This manifest also contains invalid k8s resources to simulate a release with wrong data (like one with just a comment). So I am interested in the whole file not just parts of it.

> They are not reusable. You cannot extend/modify a file since it may break an existing unit test. This lead to the growth of these kind of files which become very difficult to manage.

I agree, but that's not a problem we don't have yet. I am not even sure if future tests of that component will require the manifest at all, since they might be testing something else that is not related to the list of k8s resources of a chart. It also seemed to me like a good compromise in simplicity.

> I'd rather move this kind of manifest to a factory method (or just static vars) that returns only the piece of manifest you are interested in.

I want to know more about this. Could you add an example. Please note that I am interested in the whole file because the whole resources extraction is coupled.

```js
let manifest: IResource[] = yaml.safeLoadAll(newApp.manifest);
    // Filter out elements in the manifest that does not comply
    // with { kind: foo }
    manifest = manifest.filter(r => r && r.kind);

    const watchedKinds = ["Deployment", "Service"];
    const otherResources = manifest
      .filter(d => watchedKinds.indexOf(d.kind) < 0)
      .reduce((acc, r) => {
        // TODO: skip list resource for now
        if (r.kind === "List") {
          return acc;
        }
        acc[`${r.kind}/${r.metadata.name}`] = r;
        return acc;
      }, new Map<string, IResource>());
    this.setState({ otherResources });

    const deployments = manifest.filter(d => d.kind === "Deployment");
    const services = manifest.filter(d => d.kind === "Service");
```

Comment (andresmgot) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: > This manifest also contains invalid k8s resources to simulate a release with wrong data (like one with just a comment). So I am interested in the whole file not just parts of it.

That's mostly what I would like to avoid. You are using the same manifest to test several things. For example to test that is properly extracting a deployment and a service (for the sockets thing) but you are using the same thing for testing that it ignores invalid objects. That's difficult to discover for others (I for example didn't notice the invalid object part).

I'd split that so you have different self-contained test. If you are testing that the code is ignoring invalid objects you clearly set a manifest as simple as possible that let you test that feature.

> I want to know more about this. Could you add an example.

It doesn't have to be very smart. Maybe a simple function that return a YAML text from a object. For example for the sockets test:

```
it("sets a list of web sockets for its deployments and services", () =>
   manifest = generateManifest([
      {apiVersion: "extensions/v1beta1", kind: "Deployment", metadata: { ... }, spec: { ... }},
      {apiVersion: "v1", kind: "Service", metadata: { ... }, spec: { ... }},
   ]);
   const validProps = getValidProps(manifest);
   (the rest of the test...)
```

File: dashboard/src/components/AppView/AppView.tsx (hunk 3 above)
Comment (prydonius) [reply dashboard/src/components/AppView/AppView.tsx]: I've run into this before also, not sure why TS doesn't like `error?`, but it is specifically an issue when connecting to the Redux, I believe. Child components called from the root component loaded by the container don't have this issue in my experience.

File: dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx (hunk 2 above)
Comment (prydonius) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: I agree that it would be nicer to see more self-defined tests here rather than just a comment that spurious bogus resources aren't in the state. The generateManifest helper is cool, but how would that be used to add the empty manifests?

Comment (migmartri) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: I think we are talking about two different things here:

1 - Use a single purpose file for multiple test cases

The generator idea makes sense but the main reason for not crafting my own manifests was to test that the parser actually parses different resources separated by the `---` separator which is what we get from the API response.

I agree that it is not clear that in that manifest there are bogus data, so what I can do is to separate this into multiple files, would that work?

2 - The tests cases do too much and it is not clear what

I agree, so I'll split it.

Would these two action items work for you both?

Comment (prydonius) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: @migmartri whilst you could use a single manifest file, doesn't that make it more difficult to maintain it? For example, you can test that empty YAML is ignored with a smaller manifest that will be easier to maintain and understand what it is used for:

```
---
kind: Service
metadata:
  name: manifest-with-empty-files
...
---
# I'm empty with a comment
---
---
```

Comment (andresmgot) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: My main concern is to have tests that depends from other files when it's not very clear how. That's what I find difficult to maintain.

> The generator idea makes sense but the main reason for not crafting my own manifests was to test that the parser actually parses different resources separated by the --- separator which is what we get from the API response.

You can still add the `---` separator in the generator function (sorry that I didn't make it clear). For example (untested code):

```
function generateManifest(items) {
 let manifest = "";
 items.forEach(i => {
   manifest += "---\n" + yaml.safeDump(i);
 }
 return manifest;
```

> I agree that it is not clear that in that manifest there are bogus data, so what I can do is to separate this into multiple files, would that work?

I still don't like that very much because it's difficult to relate external files to unit tests.

> 2 - The tests cases do too much and it is not clear what
> I agree, so I'll split it.

That's true but that's not my main concern.

Comment (migmartri) [reply dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx]: Thanks @andresmgot @prydonius 

I have update the test cases by using the generator plus splitting the test cases.

Thanks for the help!

File: dashboard/src/components/AppView/AppView.test.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    });
+
+    it("does not store empty resources, bogus or without kind attribute", () => {
+      // Only one element in the otherResources array
+      // This means that no other spurious bogus (i.e missing kind: ) definitions were added.
+      const wrapper = shallow(<AppViewComponent {...validProps} />);
+      const manifest = generateYamlManifest([
+        { apiVersion: "v1", metadata: { name: "cm-one" } },
+        {},
+        "# This is a comment",
```
Comment (prydonius) [inline dashboard/src/components/AppView/AppView.test.tsx]: should we add an empty string?

File: dashboard/src/components/AppView/AppView.test.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      const otherResources: Map<string, IResource> = wrapper.state("otherResources");
+      const configMap = otherResources["ConfigMap/cm-one"];
+      expect(Object.keys(otherResources).length).toEqual(1);
+
+      expect(configMap).toBeDefined();
+      expect(configMap.metadata.name).toEqual("cm-one");
+    });
+
+    it("does not store empty resources, bogus or without kind attribute", () => {
+      // Only one element in the otherResources array
```
Comment (prydonius) [inline dashboard/src/components/AppView/AppView.test.tsx]: you mean zero?

File: dashboard/src/components/AppView/AppView.test.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        { apiVersion: "v1", metadata: { name: "cm-one" } },
+        {},
+        "# This is a comment",
+      ]);
+
+      validProps.app.manifest = manifest;
+      wrapper.setProps(validProps);
+
+      const otherResources: Map<string, IResource> = wrapper.state("otherResources");
+      expect(Object.keys(otherResources).length).toBe(0);
```
Comment (prydonius) [inline dashboard/src/components/AppView/AppView.test.tsx]: we should also ensure sockets is empty

File: dashboard/src/setupTests.ts [hunk 10]
```diff
@@ -3,5 +3,16 @@ import "raf/polyfill"; // polyfill for requestAnimationFrame
...
+const localStorageMock = {
+  clear: jest.fn(),
+  getItem: jest.fn(),
+  setItem: jest.fn(),
+};
+
+(global as any).localStorage = localStorageMock;
```
Comment (prydonius) [inline dashboard/src/setupTests.ts]: when do you use global vs window?

**Comment by migmartri** [comment]
Tests are failing but will be fixed once we merge this change https://github.com/kubeapps/kubeapps/pull/528 and rebase.

**Comment by migmartri** [comment]
@prydonius thanks for the feedback, PTAL
